app/controllers/application_controller.rb

```diff
@@ -1,10 +1,79 @@
-require './config/environment'
-
 class ApplicationController < Sinatra::Base
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if !logged_in?
+      erb :signup
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/signup' do
+    redirect '/signup' unless user_credentialed?
+    create_user
+    session[:user_id] = @user.id
+    redirect '/tweets'
+  end
+
+  get '/login' do
+    if !logged_in?
+      erb :login
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    find_user && authenticate_user
+    redirect '/login' unless user_exists?
+    session[:user_id] = @user.id
+    redirect '/tweets'
+  end
+
+  get '/logout' do
+    session.clear if logged_in?
+    redirect '/login'
+  end
+
+  helpers do
+    def current_user
+      @user = User.find(session[:user_id])
+    end
+
+    def create_user
+      @user = User.create(params)
+    end
+
+    def find_user
+      @user = User.find_by(username: params[:username])
+    end
+
+    def authenticate_user
+      @user.authenticate(params[:password])
+    end
+
+    def user_exists?
+      !!@user
+    end
+
+    def logged_in?
+      !!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```